buildengine/topological.go

```diff
@@ -8,34 +8,37 @@ import (
 
 // TopologicalSort returns a sequence of groups of modules in topological order
 // that may be built in parallel.
-func TopologicalSort(graph map[string][]string) [][]string {
-	modulesByName := map[string]bool{}
+func TopologicalSort(graph map[ProjectKey][]ProjectKey) [][]ProjectKey {
```

Comment:
Similar to above, `ProjectKey` doesn't make semantic sense here, and doesn't ensure enough type safety. I'd probably either back it out or split the types even further:

```go
//sumtype:decl
type ProjectKey interface { projectKey() }

type ModuleKey string
func (ModuleKey) projectKey() {}

type ExternalLibraryKey string
func (ExternalLibraryKey) projectKey() {}
```

This allows either to be used, or only one.

Reply:
I'll back out ProjectKey somewhat, i think that's the best thing for now.
I was also thinking along the lines of having separate types, but figured to do a fully thought through ModuleKey/ModuleName type would be a lot farther reaching, and beyond the current scope of what I wanted to change.